### PR TITLE
Add Sprout anchors to `zebra-state`

### LINKED ISSUE #2485: Add sprout anchors to `zebra-state`
Similar to #1320 

- [x] **Work on the `tree-state-sprout` branch (it already handles the state; but it has a incomplete tree implementation)**
- [ ] ~~Add sprout anchors to QueuedBlock~~
- [x] Add sprout anchors to Chain
- [x] Add sprout anchors to FinalizedState
- [x] Impl IntoDisk/From Disk for Sprout roots / note commitment tree in `disk_format.rs`
- [x] Add Sprout note commitment tree to Chain
- [x] Add Sprout note commitment tree to FinalizedState
- [x] [Regenerate stateful test disks](https://github.com/ZcashFoundation/zebra/blob/main/.github/workflows/regenerate-stateful-test-disks.yml)
- [x] Update the [test.yml](https://github.com/ZcashFoundation/zebra/blob/main/.github/workflows/test.yml) workflow to point at new stateful disk images in the same PR as the changes to state

## Notes: 

Yes Sprout is an annoying outlier in that transactions can point to a previous JoinSplit treestate _within the same transaction_ as its treestate, so we need to compute more treestates/ trees (maybe just roots, not lots of copies of trees) in order to contextually verify Sprout tx's, but this is entirely contained within the transaction verifier and then discarded. Sprout can also use the final treestates of previous blocks.

But there is no need to track interstitial Sprout treestates for:
* JoinSplits in other transactions inside the same block, or
* JoinSplits or transactions in previous blocks.

Here is the `zcashd` Sprout code:
https://github.com/zcash/zcash/blob/c34162d6ddfaa8d6276aa0628cee838a51ad9df7/src/coins.cpp#L1001-L1036

### PR DESCRIPTION
**Pull request by upbqdn**
## Motivation

In order to support Sprout, Zebra needs anchors of the Sprout note commitment tree in its state.

## Solution

This PR:
- Adds sprout anchors to Chain.
- Adds sprout anchors to FinalizedState.
- Impls IntoDisk/From Disk for Sprout roots / note commitment tree in disk_format.rs.
- Adds Sprout note commitment tree to Chain.
- Adds Sprout note commitment tree to FinalizedState.

It also:
- Increments `DATABASE_FORMAT_VERSION`.
- Updates the `test.yml` workflow to point at new stateful disk image since Zebra's finalized state now contains Sprout anchors. 

### Reviewer Checklist

  - [x] Code implements Specs and Designs
  - [x] Tests for Expected Behaviour
  - [ ] Tests for Errors

Resolves #2485.

### CODE REVIEW

File: zebra-state/src/service/non_finalized_state/chain.rs
Comment (upbqdn) [inline zebra-state/src/service/non_finalized_state/chain.rs:157]: This line causes the test `forked_equals_pushed_genesis` to fail here https://github.com/ZcashFoundation/zebra/blob/0fafa30e5d19c10e7f20085ab371ad9834274875/zebra-state/src/service/non_finalized_state/tests/prop.rs#L189

File: zebra-state/src/service/non_finalized_state/chain.rs
Comment (dconnolly) [inline zebra-state/src/service/non_finalized_state/chain.rs:1]: 🎉

File: zebra-state/src/service/non_finalized_state/chain.rs
Comment (dconnolly) [inline zebra-state/src/service/non_finalized_state/chain.rs]: The `_u` here indicates that the Sapling note commitment is specified only by the u-coordinate of the Jubjub curve point (u, v). Since Sprout note commitments are just SHA-256 hash values, we can remove the `_u`:

```suggestion
            for cm in joinsplit_data.note_commitments() {
                self.sprout_note_commitment_tree.append(*cm)?;
            }
```

**Review by dconnolly — COMMENTED** [review]
A small suggestion but looking great!

File: zebra-state/src/service/non_finalized_state/chain.rs
Comment (teor2345) [reply zebra-state/src/service/non_finalized_state/chain.rs:157]: `forked_equals_pushed_genesis` checks that you get the same chain if you:
* forking the original chain, then push some blocks, or
* push the same blocks to the original chain.

It looks like sprout anchors are implemented the same in both the fork and push code.
* Push: https://github.com/ZcashFoundation/zebra/blob/main/zebra-state/src/service/non_finalized_state.rs#L128
* Fork: https://github.com/ZcashFoundation/zebra/blob/main/zebra-state/src/service/non_finalized_state.rs#L167

So I'm wondering if there's a bug in the `NoteCommitmentTree`, `Transaction::sprout_note_commitments`, or the serialization. It's the first time a lot of that code has been tested.

You could try printing the list of block heights/hashes, transaction indexes/hashes, or note commitments that are added to `self` and `other`. That might help you find what's different.

You can stop the proptests running extra shrink tests using:
```sh
export PROPTEST_MAX_SHRINK_ITERS=0
```
That might help reduce the number of logs you have to look through.

File: zebra-state/src/service/non_finalized_state/chain.rs
Comment (upbqdn) [reply zebra-state/src/service/non_finalized_state/chain.rs]: I was wondering about the `_u`. I think the spec uses it as well, now I know.

File: zebra-state/src/service/non_finalized_state/chain.rs
Comment (conradoplg) [reply zebra-state/src/service/non_finalized_state/chain.rs:157]: Note that `eq_internal_state` checks a lot of stuff and there isn't an obvious way to tell which one failed. It could be the note commitment tree, the anchors, or the nullifiers (or some regression elsewhere).

By commenting out the Sprout checks and re-enabling one at a a time (very advanced debugging technique :laughing: ) it seems that both note commitment trees and anchors are not matching. The nullifier check is passing.

On forks, trees are recreated from the finalized state tip (in this test, it uses the `default()` of the tree for that). So I'm not sure what's going on, I second @teor2345 suggestion of printing what is being pushed to the tree and comparing to see where the difference is.

Comment (teor2345) [reply zebra-state/src/service/non_finalized_state/chain.rs:157]: It might also help to assert equality after every transaction or block is added to the chain state. That way, the test will panic on the first mismatch.

This might require rewriting the test.

Can you reproduce this issue using a test that:
1. Adds 1 block
2. Forks the chain
3. Adds 1 more block to both chains

If you can, the test would be a lot faster, and you could add transactions from the second blocks one by one.

File: zebra-state/src/service/non_finalized_state/chain.rs
Comment (daira) [reply zebra-state/src/service/non_finalized_state/chain.rs]: The spec only uses it for Sapling. For Orchard there's a similar thing with cm<sub>x</sub>, which is only the short Weierstrass x-coordinate.

File: zebra-state/src/service/non_finalized_state/tests/prop.rs
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state/tests/prop.rs]: Comment Fix: It's more than just the blocks, it's the derived data as well.

```suggestion
        // Check that the entire internal state of the forked chain corresponds to the state of
```

File: zebra-chain/src/transaction.rs
Comment (teor2345) [inline zebra-chain/src/transaction.rs]: Comment Fix: These are network upgrade names, not shielded pool names. (As of NU5, we've stopped naming network upgrades after shielded pools.)

```suggestion
    /// An NU5 transaction (`version = 5`), which supports Orchard and Sapling (but not Sprout).
```

**Review by teor2345 — COMMENTED** [review]
Looks good!

Please don't wait for me to re-review before merging, I just noticed these comments were unclear.

File: zebra-chain/src/transaction.rs
Comment (upbqdn) [reply zebra-chain/src/transaction.rs]: Thanks, I wasn't aware of that. I just wanted to make the comment consistent with the ones for other tx versions above.

File: zebra-state/src/service/non_finalized_state/tests/prop.rs
Comment (upbqdn) [reply zebra-state/src/service/non_finalized_state/tests/prop.rs]: I thought of this as well, but I couldn't come up with a suitable phrasing. This looks much better.

**Comment by upbqdn** [comment]
> Please don't wait for me to re-review before merging, I just noticed these comments were unclear.

I still need to regenerate stateful test disks. That's why this still a draft.

**Comment by upbqdn** [comment]
I updated the description of this PR.

**Comment by dconnolly** [comment]
https://github.com/ZcashFoundation/zebra/actions/runs/1517126363 ✔️ 🎉

File: zebra-chain/src/transaction.rs
Comment (dconnolly) [inline zebra-chain/src/transaction.rs]: Transaction versions are not 1:1 with network upgrades, and may be used for far longer than the network upgrade they are introduced in, so I would avoid this description:

```suggestion
    /// A `version = 5` transaction , which supports Orchard, Sapling, and transparent, but not Sprout.
```

File: zebra-state/src/service/finalized_state.rs
Comment (dconnolly) [inline zebra-state/src/service/finalized_state.rs]: Not just when `column_families` changes but any part of the disk format (eg how types are serialized/deserialized) changes

**Review by conradoplg — COMMENTED** [review]
I read this and it seems good, but haven't reviewed in detail.

File: zebra-chain/src/transaction.rs
Comment (upbqdn) [reply zebra-chain/src/transaction.rs]: Thanks, @dconnolly. There was also a similar discussion here https://github.com/ZcashFoundation/zebra/pull/3100#discussion_r757093858.

File: zebra-state/src/service/finalized_state.rs
Comment (dconnolly) [inline zebra-state/src/service/finalized_state.rs]:
```suggestion
        // Note: The [`crate::constants::DATABASE_FORMAT_VERSION`] constant must
        // be incremented each time the database format (column, serialization, etc) changes.
```

**Review by dconnolly — APPROVED** [review]
Looks good!

File: zebra-chain/src/transaction.rs
Comment (upbqdn) [reply zebra-chain/src/transaction.rs]: The reason why I edited this comment was that I wanted to make it more consistent with the ones for other transactions.

**Pull request merged**
